File: sidecar/src/kiln_sidecar/checkpoint.py
"""Pydantic mirror of the checkpoint schema.

The Rust type in `src-tauri/src/checkpoint.rs` is the other half of this
contract; both deserialize the same sample JSON in their tests. The sidecar uses
this model to validate a proposal before writing its declared decisions as
MLflow tags (Ticket 33).
"""

from __future__ import annotations

from enum import StrEnum
from typing import Final

from pydantic import BaseModel, model_validator


class Severity(StrEnum):
    CRITICAL = "critical"
    NOTABLE = "notable"
    FYI = "fyi"


class Slot(BaseModel):
    in_scope: bool
    severity: Severity
    # No `min_length` here: an out-of-scope slot may legitimately be empty/"N/A".
    # The in-scope non-empty rule (matching Rust `ProposeExperiment::validate`) is
    # enforced below, so Python accepts exactly what the Rust tool boundary does.
    answer: str

    @model_validator(mode="after")
    def _require_answer_when_in_scope(self) -> Slot:
        if self.in_scope and not self.answer.strip():
            raise ValueError("answer must be non-empty when in_scope=True")
        return self


REQUIRED_SLOTS: Final[tuple[str, ...]] = (
    "validation_strategy",
    "target_definition",
    "feature_provenance",
    "preprocessing_fit_scope",
    "data_scope_and_exclusions",
    "missing_data_handling",
    "metric_choice",
)


class ProposeExperiment(BaseModel):
    # Rust accepts any title/premise (the gate already rendered), so don't reject
    # here — Python must accept whatever passed the Rust tool boundary.
    title: str
    premise: str
    validation_strategy: Slot
    target_definition: Slot
    feature_provenance: Slot
    preprocessing_fit_scope: Slot
    data_scope_and_exclusions: Slot
    missing_data_handling: Slot
    metric_choice: Slot
    look_here: list[str]

File: src-tauri/src/checkpoint.rs
//! The "checkpoint object" — the structured `propose_experiment` payload Claude
//! fills in before the premise gate (spec §4). This Rust type and the Python
//! Pydantic mirror in `sidecar/src/kiln_sidecar/checkpoint.py` are the schema's
//! source of truth; the cross-language contract is exercised by deserializing
//! the same sample JSON in both languages' tests.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Critical,
    Notable,
    Fyi,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Slot {
    pub in_scope: bool,
    pub severity: Severity,
    /// Free-text answer or "N/A". Must be non-empty when `in_scope` is true —
    /// enforced by [`ProposeExperiment::validate`] (Ticket 31), not serde.
    pub answer: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProposeExperiment {
    pub title: String,
    pub premise: String,
    pub validation_strategy: Slot,
    pub target_definition: Slot,
    pub feature_provenance: Slot,
    pub preprocessing_fit_scope: Slot,
    pub data_scope_and_exclusions: Slot,
    pub missing_data_handling: Slot,
    pub metric_choice: Slot,
    /// Where Claude is least sure / most wants the human's eye.
    pub look_here: Vec<String>,
}

#[cfg(test)]
pub(crate) const SAMPLE_JSON: &str = r#"{
  "title": "Predict churn from the first 30 days",
  "premise": "Early engagement signals should separate churners from retained users.",
  "validation_strategy": {"in_scope": true, "severity": "critical", "answer": "Temporal split: train on cohorts before 2025-01, test after."},
  "target_definition": {"in_scope": true, "severity": "critical", "answer": "y = no login in days 31-60. No lookahead."},
  "feature_provenance": {"in_scope": true, "severity": "notable", "answer": "All features from days 0-30 only."},
  "preprocessing_fit_scope": {"in_scope": true, "severity": "notable", "answer": "Scaler fit on train split only."},
  "data_scope_and_exclusions": {"in_scope": true, "severity": "fyi", "answer": "Drop accounts active < 7 days."},
  "missing_data_handling": {"in_scope": true, "severity": "notable", "answer": "Median impute, fit on train."},
  "metric_choice": {"in_scope": true, "severity": "critical", "answer": "PR-AUC; classes are imbalanced."},
  "look_here": ["Is the 31-60 day churn window the right horizon?"]
}"#;

#[cfg(test)]
mod tests {
    use super::{ProposeExperiment, Severity, SAMPLE_JSON};

    #[test]
    fn deserializes_sample_proposal() {
        let p: ProposeExperiment = serde_json::from_str(SAMPLE_JSON).expect("deserialize sample");
        assert_eq!(p.title, "Predict churn from the first 30 days");
        assert_eq!(p.metric_choice.severity, Severity::Critical);
        assert!(p.validation_strategy.in_scope);
        assert_eq!(p.look_here.len(), 1);
    }
}
